tutor/utils.py

```diff
@@ -366,3 +366,130 @@ def format_table(rows: List[Tuple[str, ...]], separator: str = "\t") -> str:
+
+    Utilized in the `tutor local do convert-mysql-utf8mb4-charset` command
+    """
+    return f"""
+            
+            DROP PROCEDURE IF EXISTS UpdateColumns;
+            DELIMITER $$
+
+            CREATE PROCEDURE UpdateColumns()
+            BEGIN
+
+                DECLARE done_columns_loop INT DEFAULT FALSE;
+                DECLARE _table_name VARCHAR(255);
+                DECLARE _table_name_copy VARCHAR(255) DEFAULT "";
+                DECLARE _column_name VARCHAR(255);
+                DECLARE _column_type VARCHAR(255);
+                DECLARE _collation_name VARCHAR(255);
+                
+                # We explicitly upgrade the utf8mb3_general_ci collations to utf8mb4_unicode_ci
+                # The other collations are upgraded from utf8mb3_* to utf8mb4_*
+                # For any other collation, we leave it as it is
+                DECLARE columns_cur CURSOR FOR
+                SELECT 
+                    TABLE_NAME,
+                    COLUMN_NAME,
+                    COLUMN_TYPE,
+                    CASE 
+                        WHEN COLLATION_NAME LIKE CONCAT('{charset_to_upgrade_from}', '_general_ci') THEN 'utf8mb4_unicode_ci'
+                        WHEN COLLATION_NAME LIKE CONCAT('{charset_to_upgrade_from}', '_%') THEN CONCAT('{charset}', SUBSTRING_INDEX(COLLATION_NAME, '{charset_to_upgrade_from}', -1))
+                        ELSE COLLATION_NAME 
+                    END AS COLLATION_NAME 
+                FROM 
+                    INFORMATION_SCHEMA.COLUMNS
+                WHERE 
+                    TABLE_SCHEMA = '{database}'
+                    AND COLLATION_NAME IS NOT NULL {query_to_append};
+                DECLARE CONTINUE HANDLER FOR NOT FOUND SET done_columns_loop = TRUE;
+                OPEN columns_cur;
+                columns_loop: LOOP
+                    FETCH columns_cur INTO _table_name, _column_name, _column_type, _collation_name;
+                    
+                    IF done_columns_loop THEN
+                    LEAVE columns_loop;
+                    END IF;
+                    
+                    # First, upgrade the default charset and collation of the table
+                    If _table_name <> _table_name_copy THEN
+                    select _table_name;
+                    SET FOREIGN_KEY_CHECKS = 0;
+                    SET @stmt = CONCAT('ALTER TABLE `', _table_name, '` CONVERT TO CHARACTER SET {charset} COLLATE {collation};'); 
+                    PREPARE query FROM @stmt;
+                    EXECUTE query;
+                    DEALLOCATE PREPARE query; 
+                    SET FOREIGN_KEY_CHECKS = 1;
+                    SET _table_name_copy = _table_name;
+                    END IF;
+                    
+                    # Then, upgrade the default charset and collation of each column
+                    # This sequence of table -> column is necessary to preserve column defaults
```

Comment:
IF we are updating table again in UpdateTables, how will it not cause any issues?

Reply:
UpdateTables will only update the tables that have the default collation of something other than `utf8mb4_*`, i.e, tables that weren't updated in UpdateColumns.

Reference: https://github.com/overhangio/tutor/pull/1079/files#diff-46b7a04b09b0ffeb02b4f36ed494cbed6eaa7ea1efb4c46f9951b4ff6b0a3a37R464